timemachine/fe/free_energy.py

```diff
@@ -268,6 +278,41 @@ def frames(self) -> List[StoredArrays]:
     def boxes(self) -> List[NDArray]:
         return [np.array(traj.boxes) for traj in self.trajectories]
 
+    def compute_u_kn(self) -> Tuple[NDArray, NDArray]:
+        """get MBAR input matrices u_kn and N_k"""
+
+        # TODO: there's probably a much faster idiom for this -- shouldn't take a whole minute...
+        N_k = [len(traj.frames) for traj in self.trajectories]
+        N, K = sum(N_k), len(N_k)
+        assert len(self.final_result.initial_states) == K
+
+        kBTs = [BOLTZ * state.integrator.temperature for state in self.final_result.initial_states]
+
+        # assume all states are compatible
+        s_0 = self.final_result.initial_states[0]
+        for s in self.final_result.initial_states[1:]:
+            assert_ensembles_compatible(s_0, s)
+            assert_potentials_compatible(s_0.potentials, s.potentials)
+
+        u_kn = np.zeros((K, N))
+        for k, initial_state in enumerate(self.final_result.initial_states):
+            kBT = kBTs[k]
+
+            bound_impl = initial_state.to_bound_impl()
+
+            def u(x, box):
+                # TODO: bound_impl.execute_batch
```

Comment:
Could this actually use the unbound `execute_batch`, and batch over both coordinates and parameters? (OK with postponing optimizations like this especially if they require significant refactoring)

Reply:
Applied in https://github.com/proteneer/timemachine/pull/1387/commits/978b1da1f6974263f0cfc8ce8b5a06a9293e1a1c  (using `unbound_impl.execute_batch_sparse`), did not require significant refactoring